feat: add additional operations to `BaseFlags`

celsiusnarhwal opened the pull request:
## Summary

This pull request implements addition (+), subtraction (-), union (|), intersection (&), and inversion (~) operations on the `BaseFlags` class and accordingly updates the documentation for its inheritors. The addition and union operations are equivalent and their operators interchangeable.

I made this change with the `Permissions` class in mind, intending to enable functionality such as adding two `Permissions` objects together to obtain the combined set of permissions they represent; subtracting two `Permissions` objects to set all permissions that are `True` in one object to `False` in the other; finding the intersection of two `Permissions` objects to obtain the set of permissions that are `True` in both; and inverting a `Permissions` object to set all of its `True` permissions to `False` and vice versa. However, this PR enables these operations on _all_ inheritors of `BaseFlags` - the previously-described actions could be similarly applied to objects of the `Intents` class, for example. 

The addition, subtraction, union, and intersection operations can operate on a `BaseFlags` object and a) another `BaseFlags` object, or b) a `flag_value` object. Reverse methods for each of these operations are also implemented, allowing operations between `BaseFlags` and `flag_value` objects to be properly commutative.

## Information

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...).

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.

EmreTech commented:
Considered adding this to all flags in `flag.py`?

plun1331 commented on discord/permissions.py:
```diff
@@ -149,11 +161,58 @@ def is_strict_superset(self, other: Permissions) -> bool:
+            raise TypeError(f"'-' not supported between instances of '{type(self)}' and '{type(other)}'")
+
+    def __invert__(self):
+        if isinstance(self, Permissions):
```
Unnecessary

plun1331 commented on discord/permissions.py:
```diff
@@ -149,11 +161,58 @@ def is_strict_superset(self, other: Permissions) -> bool:
+        else:
+            raise TypeError(f"'~' not supported for '{type(self)}'")
+
+    __rand__ = __and__
+    __ror__ = __or__
+    __radd__ = __add__
+    __rsub__ = __sub__
```
These should be typehinted probably

plun1331 commented on discord/permissions.py:
```diff
@@ -149,11 +161,58 @@ def is_strict_superset(self, other: Permissions) -> bool:
     __lt__: Callable[[Permissions], bool] = is_strict_subset
     __gt__: Callable[[Permissions], bool] = is_strict_superset
 
+
```
These extra newlines shouldn't be here

plun1331 reviewed (changes requested):
While this is a nice feature, implementing it in the [`BaseFlags`](https://github.com/Pycord-Development/pycord/blob/1bb5ac38d20ebb6a84eb348cb2b6edb4d6e2f4cc/discord/flags.py#L102) class would probably be a better solution, also allowing you to perform the same operations on everything that inherits from it. (intents, etc)

EmreTech commented:
> While this is a nice feature, implementing it in the [`BaseFlags`](https://github.com/Pycord-Development/pycord/blob/1bb5ac38d20ebb6a84eb348cb2b6edb4d6e2f4cc/discord/flags.py#L102) class would probably be a better solution, also allowing you to perform the same operations on everything that inherits from it. (intents, etc)

Yeah this is what I was trying to say

Lulalaby reviewed (dismissed):
That looks like a mess

celsiusnarhwal commented:
> While this is a nice feature, implementing it in the [`BaseFlags`](https://github.com/Pycord-Development/pycord/blob/1bb5ac38d20ebb6a84eb348cb2b6edb4d6e2f4cc/discord/flags.py#L102) class would probably be a better solution, also allowing you to perform the same operations on everything that inherits from it. (intents, etc)

That actually is a better idea; I made this PR with only the `Permissions` class in mind, so that did not initially occur to me. Nevertheless, I have moved my code changes to the `BaseFlags` class (with typehints this time) and updated the documentation for its inheritors accordingly.

EmreTech commented on discord/flags.py:
```diff
@@ -39,6 +39,7 @@
     overload,
 )
 
+import discord
```
Why are you directly importing the `discord` namespace? Just import what you need

EmreTech commented on discord/flags.py:
```diff
@@ -140,6 +141,44 @@ def __iter__(self) -> Iterator[Tuple[str, bool]]:
+    __rand__: Callable[[BaseFlags], bool] = __and__
+    __ror__: Callable[[BaseFlags], bool] = __or__
+    __radd__: Callable[[BaseFlags], bool] = __add__
+    __rsub__: Callable[[BaseFlags], bool] = __sub__
```
I'm a little confused with these type annotations. The `__and__` and `__or__` functions support BaseFlags and flag_value for the `other` parameter

Lulalaby replied:
Same confusion here

EmreTech commented on discord/flags.py:
```diff
@@ -177,6 +216,19 @@ class SystemChannelFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+            Adds two flags together.
```
The comment should go below the `.. describe::` part

EmreTech commented on discord/flags.py:
```diff
@@ -250,6 +302,19 @@ class MessageFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+            Adds two flags together.
```
See above

EmreTech commented on discord/flags.py:
```diff
@@ -490,6 +568,19 @@ class Intents(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+            Adds two flags together.
```
See above

EmreTech commented on discord/flags.py:
```diff
@@ -1030,6 +1121,18 @@ class MemberCacheFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+            Adds two flags together.
```
See above

EmreTech commented on discord/flags.py:
```diff
@@ -1155,6 +1258,18 @@ class ApplicationFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+            Adds two flags together.
```
See above

EmreTech commented on discord/flags.py:
```diff
@@ -1253,6 +1368,18 @@ class ChannelFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+            Adds two flags together.
```
See above

EmreTech commented on discord/permissions.py:
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
              Checks if a permission is a strict subset of another permission.
         .. describe:: x > y
 
+            Adds two permissions together.
```
See above

EmreTech reviewed (commented):
Looks good otherwise

EmreTech commented on discord/flags.py:
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
+         Adds two flags together.
+
+        .. describe:: x - y
```
Newline + indent for the description

Lulalaby replied:
```suggestion
        .. describe:: x + y
        Adds two flags together.

        .. describe:: x - y
```

EmreTech commented on discord/flags.py:
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
+        .. describe:: x - y
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
+        .. describe:: x & y
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
+        .. describe:: x + y
+         Adds two flags together.
+
+        .. describe:: x - y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
+        .. describe:: x - y
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
+        .. describe:: x & y
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
+        .. describe:: x + y
+         Adds two flags together.
+
+        .. describe:: x - y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
+        .. describe:: x - y
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
+        .. describe:: x & y
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
+        .. describe:: x + y
+         Adds two flags together.
+
+        .. describe:: x - y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
+        .. describe:: x - y
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
+        .. describe:: x & y
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1030,6 +1120,18 @@ class MemberCacheFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Why is this here for the documentation of a specific flag

plun1331 replied:
It's under `MemberCacheFlags`

Lulalaby replied:
Shouldn't _MemberCacheFlags(BaseFlags)_ get the operators / methods from base flags already?

plun1331 replied:
> Shouldn't _MemberCacheFlags(BaseFlags)_ get the operators / methods from base flags already?

It'll get them, they just won't be in the documentation unless they're put there

Lulalaby replied:
Wouldn't autodoc be a good approach? Or doesn't it work.

plun1331 replied:
autodoc is for detecting modules/classes and their methods and automatically adding them to the documentation, it won't merge docstrings from two different classes

Lulalaby replied:
What the.. that sounds so useless. 
Well ok fair.

EmreTech commented on discord/flags.py:
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
+        .. describe:: x + y
+         Adds two flags together.
+
+        .. describe:: x - y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
+        .. describe:: x - y
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
+        .. describe:: x & y
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
+        .. describe:: x + y
+         Adds two flags together.
+
+        .. describe:: x - y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
+        .. describe:: x - y
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
+        .. describe:: x & y
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Newline + indent for the description

EmreTech commented on discord/permissions.py:
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
              Checks if a permission is a strict subset of another permission.
         .. describe:: x > y
 
+         .. describe:: x + y
```
Newline + indent for the description

EmreTech commented on discord/permissions.py:
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
              Checks if a permission is a strict subset of another permission.
         .. describe:: x > y
 
+         .. describe:: x + y
+         Adds two permissions together.
+
+        .. describe:: x - y
```
Newline + indent for the description

Lulalaby replied:
```suggestion
        .. describe:: x + y
        Adds two permissions together.

        .. describe:: x - y
```

EmreTech commented on discord/permissions.py:
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
+        .. describe:: x - y
+        Subtracts two permissions from each other.
+
+        .. describe:: x & y
```
Newline + indent for the description

EmreTech commented on discord/permissions.py:
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
+        .. describe:: x & y
+        Returns the intersection of two permissions.
+
+        .. describe:: ~x
```
Newline + indent for the description

EmreTech commented on discord/flags.py:
```diff
@@ -140,6 +141,44 @@ def __iter__(self) -> Iterator[Tuple[str, bool]]:
+        else:
+            raise TypeError(f"'&' not supported between instances of {type(self)} and {type(other)}")
+
+    def __or__(self, other):
```
This doesn't seem to be documented anywhere

Lulalaby commented on discord/flags.py:
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
+         Adds two flags together.
```
```suggestion
        Adds two flags together.
```

EmreTech reviewed (commented):
Sorry for the repetition, but it looks fine otherwise

Lulalaby commented on discord/flags.py:
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
             Note that aliases are not shown.
 
+        .. describe:: x + y
+         Adds two flags together.
```
```suggestion
        Adds two flags together.
```

Lulalaby commented on discord/flags.py:
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
+         Adds two flags together.
```
```suggestion
        Adds two flags together.
```

Lulalaby commented on discord/flags.py:
```diff
@@ -1030,6 +1120,18 @@ class MemberCacheFlags(BaseFlags):
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
+         Adds two flags together.
```
```suggestion
        Adds two flags together.
```

Lulalaby commented on discord/flags.py:
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
             Note that aliases are not shown.
 
+        .. describe:: x + y
+         Adds two flags together.
```
```suggestion
        Adds two flags together.
```

Lulalaby commented on discord/flags.py:
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
             Note that aliases are not shown.
 
+        .. describe:: x + y
+         Adds two flags together.
```
```suggestion
        Adds two flags together.
```
